Source/DafnyCore/Compilers/Rust/Dafny-compiler-rust-proofs.dfy

```diff
@@ -0,0 +1,362 @@
+      if s[0] == '_' {
+      } else if s[0] == '?' {
+      } else if s[0] == '\'' {
+      } else if s[0] == '#' {
+      } else {
+      }
+    }
+  }
+
+  lemma dafnyEscapeCharacteristic(s: string)
+    requires IsDafnyIdTail(s)
+    ensures isDafnyEncodedIdTail(dafnyEscape(s))
+  {
+    if |s| == 0 {
+    } else {
+      dafnyEscapeCharacteristic(s[1..]);
+      if s[0] == '_' {
+      } else if s[0] == '?' {
+      } else if s[0] == '\'' {
+      } else if s[0] == '#' {
+      } else {
+      }
+    }
+  }
+
+  function reverseReplaceDots(i: string): string {
+    if |i| <= 1 then i
+    else if i[0] == '_' then
+      if i[1] == 'd' then "." + reverseReplaceDots(i[2..])
+      else [i[0]] + [i[1]] + reverseReplaceDots(i[2..])
+    else
+      [i[0]] + reverseReplaceDots(i[1..])
+  }
+
+  predicate isDafnyEncodedIdTail(i: string) {
+    if |i| == 0 then true
+    else
+    if i[0] == '_' then
+      |i| >= 2 &&
+      i[1] in "_qkh" && isDafnyEncodedIdTail(i[2..])
+    else
+      i[0] in "aqkhd." &&
+      isDafnyEncodedIdTail(i[1..])
+  }
+
+  predicate isDafnyEncodedId(i: string) {
+    if |i| == 0 then true
+    else i[0] in "aqkhd" && isDafnyEncodedIdTail(i[1..])
+  }
+
+  lemma dafnyEscapeCorrect(s: string)
+    requires IsDafnyIdTail(s)
+    ensures isDafnyEncodedIdTail(dafnyEscape(s))
+            && (IsDafnyId(s) && |s| > 0 ==> dafnyEscape(s)[0] in "aqkhd")
+  {
+    if |s| == 0 {
+    } else {
+      if s[0] == '_' { // "__" + dafnyEscape(s[1..])
+      } else if s[0] == '?' { // "_q" + dafnyEscape(s[1..]) else
+      } else if s[0] == '\'' { // "_k" + dafnyEscape(s[1..]) else
+      } else if s[0] == '#' { // "_h" + dafnyEscape(s[1..]) else=
+      } else { // [s[0]] + dafnyEscape(s[1..])=
+        dafnyEscapeCorrect(s[1..]);
+      }
+    }
+  }
+  // replaceDots is not reversible generally speaking.
+  // Indeed, both _d and . get translated to _d.
+  lemma replaceDotsNotReversible()
+    ensures replaceDots("_d") == replaceDots(".")
+    ensures !isDafnyEncodedIdTail("_d")
+  {}
+
+  // However, if we restrict ourself to a Dafny encoding, it's reversible
+  lemma {:rlimit 500}
+  replaceDotsReversible(i: string)
+    requires isDafnyEncodedIdTail(i)
+    ensures reverseReplaceDots(replaceDots(i)) == i
+  {
+    if |i| == 0 {
+    } else if i[0] == '.' { //"_d" + replaceDots(i[1..])
+      replaceDotsReversible(i[1..]);
+      assert reverseReplaceDots(replaceDots(i)) == i;
+    } else { // [i[0]] + replaceDots(i[1..])
+      assert replaceDots(i) == [i[0]] + replaceDots(i[1..]);
+      if i[0] == '_' {
+        assert |i| >= 2 &&
+               i[1] in "_qkh" && isDafnyEncodedIdTail(i[2..]);
+        replaceDotsReversible(i[2..]);
+        assert reverseReplaceDots(replaceDots(i)) == i;
+      } else {
+        replaceDotsReversible(i[1..]);
+        assert reverseReplaceDots(replaceDots(i)) == i;
+      }
+    }
+  }
+
+  // The reason why escapeIndent is correct is that
+  // - It receives either
```

Comment:
This sounds like a big old implicit precondition we should make explicit (since it's defined in Dafny)

Reply:
You mean, adding this as a precondition to the function `escapeIdent` itself?
Since the reception is total ("anything else" is the last case), the precondition of `escapeIdent` is actually true all the time.
Without this being an objection, the disjunction in the precondition of that function is there only to figure out a particular type of bijection (I love writing this -ctions :-) )
so writing a precondition would make sense if I split the function "escapeIdent" itself, which at this point is not necessary. Unless you have another idea to format it?

I also renamed "escapeIndent" to "escapeIdent" in the next PR since it's a typo.

Reply:
Oh I missed the "everything else" case and therefore thought there was a non-trivial precondition, ignore :)